src/Libraries/Microsoft.Extensions.TimeProvider.Testing/FakeTimeProvider.cs

```diff
@@ -158,28 +159,43 @@ internal void RemoveWaiter(FakeTimeProviderTimer.Waiter waiter)
         }
     }
 
-    private List<FakeTimeProviderTimer.Waiter> GetWaitersToWake()
+    private FakeTimeProviderTimer.Waiter? TryGetWaiterToWake(DateTimeOffset targetNow)
     {
-        var l = new List<FakeTimeProviderTimer.Waiter>(Waiters.Count);
-        foreach (var w in Waiters)
+        var candidate = default(FakeTimeProviderTimer.Waiter);
+
+        lock (Waiters)
         {
-            if (_now >= w.WakeupTime)
+            if (Waiters.Count == 0)
             {
-                l.Add(w);
+                return null;
             }
-        }
 
-        return l;
-    }
-
-    private void WakeWaiters(List<FakeTimeProviderTimer.Waiter> waiters)
-    {
-        foreach (var w in waiters)
-        {
-            if (_now >= w.WakeupTime)
+            // Search through Waiters and find the waiter with the minimum
+            // WakeupTime which is less than or equal to targetNow.
+            foreach (var waiter in Waiters)
             {
-                w.TriggerAndSchedule(false);
+                if (waiter.WakeupTime > targetNow)
+                {
+                    continue;
+                }
+
+                if (candidate is null)
+                {
+                    candidate = waiter;
+                    continue;
+                }
+
+                // This finds the waiter with the minimum WakeupTime
+                // and also ensures that if multiple waiters have the same
+                // the one that is picked is also the one that was scheduled
+                // first.
```

Comment:
[nit] Developers are generally constrained in the vertical plane more than in the horizontal plane. We found in few other repos that a line ~120-150 char long facilitates a good developer experience.

Reply:
My interpretation of your suggestion here was that I should not break the comment into so many lines. Ive changed that. Let me now if I misunderstood your intentions.